File: src/models/compression.ts
import type { Slice, MergedSlice } from './types'

function token_distance(a: MergedSlice, b: MergedSlice): number {
  let d = 0
  if (a.state !== b.state) {
    d += 4
  } else if (a.state === 'Uninterruptible Sleep' && a.io_wait !== b.io_wait) {
    d += 2
  }
  if (a.name !== b.name) {
    d += (a.name === null || b.name === null) ? 1 : 2
  }
  if (a.blocked_function !== b.blocked_function) {
    d += (a.blocked_function === null || b.blocked_function === null) ? 0.5 : 1
  }
  return d
}

function merge_cost(a: MergedSlice, b: MergedSlice): number {
  const dist = token_distance(a, b)
  if (dist === 0) {
    return 0.01 * Math.log1p((a.dur + b.dur) / 1e6)
  }
  const loser = a.dur <= b.dur ? a : b
  const loser_weight = Math.log1p(loser.dur / 1e6)
    * (1 + (loser.name !== null ? 1 : 0) + (loser.blocked_function !== null ? 1 : 0))
  return dist * loser_weight
}

function merge_two(a: MergedSlice, b: MergedSlice): MergedSlice {
  const winner = (a.dur >= b.dur) ? a : b
  return {
    ts: a.ts,
    tsRel: a.tsRel,
    dur: a.dur + b.dur,
    state: winner.state,
    io_wait: winner.io_wait,
    name: winner.name,
    depth: winner.depth,
    blocked_function: winner.blocked_function,
    _merged: (a._merged || 1) + (b._merged || 1),
  }
}

export function build_merge_cache(raw_data: Slice[]): Map<number, MergedSlice[]> {
  const cache = new Map<number, MergedSlice[]>()
  if (raw_data.length === 0) return cache

  let seq: MergedSlice[] = raw_data.map(d => ({
    ...d, tsRel: d.ts - raw_data[0].ts, _merged: 1,
  }))
  cache.set(seq.length, seq.map(d => ({ ...d })))

  while (seq.length > 2) {
    let best_i = 0, best_cost = Infinity
    for (let i = 0; i < seq.length - 1; i++) {
      const c = merge_cost(seq[i], seq[i + 1])
      if (c < best_cost) { best_cost = c; best_i = i }
    }
    seq = [
      ...seq.slice(0, best_i),
      merge_two(seq[best_i], seq[best_i + 1]),
      ...seq.slice(best_i + 2),
    ]
    if (!cache.has(seq.length)) cache.set(seq.length, seq.map(d => ({ ...d })))
  }
  return cache
}

export function get_compressed(cache: Map<number, MergedSlice[]>, orig_n: number, target: number): MergedSlice[] {
  let best = orig_n
  for (const k of cache.keys()) if (k >= target && k < best) best = k
  return cache.get(best) ?? cache.get(2) ?? []
}

File: android/app/src/main/java/com/swiperf/app/data/compression/Compression.kt
package com.swiperf.app.data.compression

import com.swiperf.app.data.model.MergedSlice
import com.swiperf.app.data.model.Slice
import kotlin.math.ln1p
import kotlin.math.max

object Compression {

    private fun tokenDistance(a: MergedSlice, b: MergedSlice): Double {
        var d = 0.0
        if (a.state != b.state) {
            d += 4.0
        } else if (a.state == "Uninterruptible Sleep" && a.ioWait != b.ioWait) {
            d += 2.0
        }
        if (a.name != b.name) {
            d += if (a.name == null || b.name == null) 1.0 else 2.0
        }
        if (a.blockedFunction != b.blockedFunction) {
            d += if (a.blockedFunction == null || b.blockedFunction == null) 0.5 else 1.0
        }
        return d
    }

    private fun mergeCost(a: MergedSlice, b: MergedSlice): Double {
        val dist = tokenDistance(a, b)
        if (dist == 0.0) {
            return 0.01 * ln1p((a.dur + b.dur) / 1e6)
        }
        val loser = if (a.dur <= b.dur) a else b
        val loserWeight = ln1p(loser.dur / 1e6) *
            (1 + (if (loser.name != null) 1 else 0) + (if (loser.blockedFunction != null) 1 else 0))
        return dist * loserWeight
    }

    private fun mergeTwo(a: MergedSlice, b: MergedSlice): MergedSlice {
        val winner = if (a.dur >= b.dur) a else b
        return MergedSlice(
            ts = a.ts,
            tsRel = a.tsRel,
            dur = a.dur + b.dur,
            state = winner.state,
            ioWait = winner.ioWait,
            name = winner.name,
            depth = winner.depth,
            blockedFunction = winner.blockedFunction,
            merged = a.merged + b.merged
        )
    }

    fun buildMergeCache(rawData: List<Slice>): Map<Int, List<MergedSlice>> {
        val cache = mutableMapOf<Int, List<MergedSlice>>()
        if (rawData.isEmpty()) return cache

        val baseTs = rawData[0].ts
        var seq = rawData.map { d ->
            MergedSlice(
                ts = d.ts, tsRel = d.ts - baseTs, dur = d.dur,
                state = d.state, ioWait = d.ioWait, name = d.name,
                depth = d.depth, blockedFunction = d.blockedFunction, merged = 1
            )
        }.toMutableList()
        cache[seq.size] = seq.toList()

        while (seq.size > 2) {
            var bestI = 0
            var bestCost = Double.MAX_VALUE
            for (i in 0 until seq.size - 1) {
                val c = mergeCost(seq[i], seq[i + 1])
                if (c < bestCost) { bestCost = c; bestI = i }
            }
            val merged = mergeTwo(seq[bestI], seq[bestI + 1])
            seq.removeAt(bestI + 1)
            seq[bestI] = merged
            if (!cache.containsKey(seq.size)) {
                cache[seq.size] = seq.toList()
            }
        }
        return cache
    }

    fun getCompressed(cache: Map<Int, List<MergedSlice>>, origN: Int, target: Int): List<MergedSlice> {
        val t = max(2, target)
        var best = origN
        for (k in cache.keys) {
            if (k >= t && k < best) best = k
        }
        return cache[best] ?: cache[2] ?: emptyList()
    }
}
